[HDR] Enable HDR display for SVG <image> elements

shallawa opened the pull request:
#### afbdfa6e4029e7ff7d76e8b04d049915f4528dad
<pre>
[HDR] Enable HDR display for SVG &lt;image&gt; elements
<a href="https://bugs.webkit.org/show_bug.cgi?id=289419#">https://bugs.webkit.org/show_bug.cgi?id=289419#</a>
<a href="https://rdar.apple.com/146582114">rdar://146582114</a>

Reviewed by Cameron McCormack.

BitmapImage and SVGImage will implement the virtual method hasPaintedHDRContent().
This will enable HDR display support for these images. This method will replace
the existing method headroom() since it does not make sense to have SVGImage
implement it. Similarly setHasPaintedHDRContentForTesting() will replace
setHeadroomForTesting().

LegacyRenderSVGImage needs to implement notifyFinished() which will be used to
mark the document to have setHasPaintedHDRContent(). It will be used also to call
RenderLayer::contentChanged(ContentChangeType::Image) for the SVGImage case.

To go from LegacyRenderSVGImage inside an SVGImage to the parent layer which owns
the SVGImage, the SVGImageChromeClient will be used. A new virtual method called
didFinishLoadingImageForSVGImage() will be added to ChromeClient. This method will
be called form LegacyRenderSVGImage::notifyFinished() through Page. This method
will use the ImageObserver to call the method RenderElement::imageContentChanged()
which will mark the layer for image content change.

* LayoutTests/TestExpectations:
* LayoutTests/compositing/hdr/hdr-basic-image.html:
* LayoutTests/compositing/hdr/hdr-css-image.html:
* LayoutTests/compositing/hdr/hdr-svg-inline-image.html: Copied from LayoutTests/compositing/hdr/hdr-basic-image.html.
* LayoutTests/fast/images/hdr-basic-image.html:
* LayoutTests/fast/images/hdr-css-image.html:
* LayoutTests/fast/images/hdr-svg-inline-image-expected.html: Added.
* LayoutTests/fast/images/hdr-svg-inline-image.html: Copied from LayoutTests/fast/images/hdr-basic-image.html.
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/ios/compositing/hdr/hdr-svg-inline-image-expected.txt: Added.
* LayoutTests/platform/mac-wk2/TestExpectations:
* LayoutTests/platform/mac-wk2/compositing/hdr/hdr-svg-inline-image-expected.txt: Added.
* Source/WebCore/html/ImageBitmap.cpp:
* Source/WebCore/loader/cache/CachedImage.cpp:
(WebCore::CachedImage::hasPaintedHDRContent const):
(WebCore::CachedImage::CachedImageObserver::imageContentChanged):
(WebCore::CachedImage::imageContentChanged):
(WebCore::CachedImage::headroom const): Deleted.
* Source/WebCore/loader/cache/CachedImage.h:
* Source/WebCore/loader/cache/CachedImageClient.h:
(WebCore::CachedImageClient::imageContentChanged):
* Source/WebCore/page/ChromeClient.h:
(WebCore::ChromeClient::didFinishLoadingImageForSVGImage):
* Source/WebCore/page/Page.cpp:
(WebCore::Page::didFinishLoadingImageForSVGImage):
* Source/WebCore/page/Page.h:
* Source/WebCore/platform/graphics/BitmapImage.cpp:
(WebCore::BitmapImage::draw):
* Source/WebCore/platform/graphics/BitmapImage.h:
* Source/WebCore/platform/graphics/BitmapImageSource.h:
* Source/WebCore/platform/graphics/Image.h:
(WebCore::Image::hasPaintedHDRContent const):
(WebCore::Image::headroom const): Deleted.
* Source/WebCore/platform/graphics/ImageObserver.h:
* Source/WebCore/platform/graphics/ImageSource.h:
(WebCore::ImageSource::setHasPaintedHDRContentForTesting):
(WebCore::ImageSource::hasPaintedHDRContentForTesting const):
(WebCore::ImageSource::setHeadroomForTesting): Deleted.
(WebCore::ImageSource::headroomForTesting const): Deleted.
* Source/WebCore/rendering/RenderElement.cpp:
(WebCore::RenderElement::imageContentChanged):
* Source/WebCore/rendering/RenderElement.h:
* Source/WebCore/rendering/RenderImage.cpp:
(WebCore::RenderImage::notifyFinished):
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::calculateClipRects const):
* Source/WebCore/rendering/RenderLayer.h:
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::PaintedContentsInfo::PaintedContentsInfo):
(WebCore::PaintedContentsInfo::isContentsTypeSatisfied const):
(WebCore::PaintedContentsInfo::isRenderElementWithHDR):
(WebCore::PaintedContentsInfo::determineContentsType):
(WebCore::RenderLayerBacking::updateDrawsContent):
(WebCore::RenderLayerBacking::isRenderElementWithHDR const):
(WebCore::PaintedContentsInfo::isReplacedElementWithHDR): Deleted.
(WebCore::RenderLayerBacking::isReplacedElementWithHDR const): Deleted.
* Source/WebCore/rendering/RenderLayerBacking.h:
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGImage.cpp:
(WebCore::LegacyRenderSVGImage::notifyFinished):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGImage.h:
* Source/WebCore/svg/SVGImageElement.h:
* Source/WebCore/svg/graphics/SVGImage.cpp:
(WebCore::SVGImage::hasPaintedHDRContent const):
* Source/WebCore/svg/graphics/SVGImage.h:
* Source/WebCore/svg/graphics/SVGImageClients.h:
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::setHasPaintedHDRContentForTesting):
(WebCore::Internals::setHeadroomForTesting): Deleted.
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:
* Tools/TestWebKitAPI/Tests/WebCore/SVGImageCasts.cpp:

Canonical link: <a href="https://commits.webkit.org/291982@main">https://commits.webkit.org/291982@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b3159641d1917a85a400d13526e8bbac803fe66b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/94600 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/14192 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/3995 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/99621 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/45112 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/96650 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/14487 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/22620 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/72183 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/29489 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/97602 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/10788 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/85411 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/52514 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/10481 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/3113 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/44437 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/80697 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/3217 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/101663 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/21651 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/15794 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/81177 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/21899 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/81434 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/80555 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/20089 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/25115 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/2504 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/14867 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/21629 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/26752 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/21302 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/24767 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/23040 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

heycam commented on Source/WebCore/rendering/svg/legacy/LegacyRenderSVGImage.cpp:
```diff
@@ -65,6 +65,25 @@ CheckedRef<RenderImageResource> LegacyRenderSVGImage::checkedImageResource() con
 
+void LegacyRenderSVGImage::notifyFinished(CachedResource& newImage, const NetworkLoadMetrics& metrics, LoadWillContinueInAnotherProcess loadWillContinueInAnotherProcess)
+{
+    if (renderTreeBeingDestroyed())
+        return;
```
Is it correct to skip the call to `LegacyRenderSVGModelObject::notifyFinished` in this case?

shallawa replied:
I think yes. I copied this condition from RenderImage::imageChanged() which is called from CachedImage::notifyObservers(). This calling sequence is very similar to what I am trying to do when an `<image>` element inside an `<img src="foo.svg">` finishes loading.

heycam commented on Source/WebCore/rendering/RenderLayer.h:
```diff
@@ -596,7 +596,7 @@ class RenderLayer final : public CanMakeSingleThreadWeakPtr<RenderLayer>, public
     void determineNonLayerDescendantsPaintedContent(PaintedContentRequest&) const;
 #if HAVE(SUPPORT_HDR_DISPLAY)
     // True of if renderer itself draws HDR content, no traversal is done.
```
Nit: Existing typo; s/True of if/True if/?

heycam commented on LayoutTests/fast/images/hdr-svg-inline-image.html:
```diff
@@ -0,0 +1,38 @@
+        var image = new Image;
+        image.onload = (() => {
+            if (window.internals)
+                internals.setHasPaintedHDRContentForTesting(image, 5);
```
The `5` argument isn't used.

heycam commented on LayoutTests/fast/images/hdr-css-image.html:
```diff
@@ -42,7 +42,7 @@
                     image.onload = (e) => {
                         if (window.internals)
-                            internals.setHeadroomForTesting(image, 5);
+                            internals.setHasPaintedHDRContentForTesting(image,);
```
Trailing comma.

heycam commented on LayoutTests/fast/images/hdr-basic-image.html:
```diff
@@ -25,7 +25,7 @@
         image.onload = (() => {
             if (window.internals)
-                internals.setHeadroomForTesting(image, 5);
+                internals.setHasPaintedHDRContentForTesting(image, 5);
```
The `5` argument isn't used.

heycam reviewed (approved):
r=me with some comments.

shallawa committed:
[HDR] Enable HDR display for SVG <image> elements
https://bugs.webkit.org/show_bug.cgi?id=289419#
rdar://146582114

Reviewed by Cameron McCormack.

BitmapImage and SVGImage will implement the virtual method hasPaintedHDRContent().
This will enable HDR display support for these images. This method will replace
the existing method headroom() since it does not make sense to have SVGImage
implement it. Similarly setHasPaintedHDRContentForTesting() will replace
setHeadroomForTesting().

LegacyRenderSVGImage needs to implement notifyFinished() which will be used to
mark the document to have setHasPaintedHDRContent(). It will be used also to call
RenderLayer::contentChanged(ContentChangeType::Image) for the SVGImage case.

To go from LegacyRenderSVGImage inside an SVGImage to the parent layer which owns
the SVGImage, the SVGImageChromeClient will be used. A new virtual method called
didFinishLoadingImageForSVGImage() will be added to ChromeClient. This method will
be called form LegacyRenderSVGImage::notifyFinished() through Page. This method
will use the ImageObserver to call the method RenderElement::imageContentChanged()
which will mark the layer for image content change.

* LayoutTests/TestExpectations:
* LayoutTests/compositing/hdr/hdr-basic-image.html:
* LayoutTests/compositing/hdr/hdr-css-image.html:
* LayoutTests/compositing/hdr/hdr-svg-inline-image.html: Copied from LayoutTests/compositing/hdr/hdr-basic-image.html.
* LayoutTests/fast/images/hdr-basic-image.html:
* LayoutTests/fast/images/hdr-css-image.html:
* LayoutTests/fast/images/hdr-svg-inline-image-expected.html: Added.
* LayoutTests/fast/images/hdr-svg-inline-image.html: Copied from LayoutTests/fast/images/hdr-basic-image.html.
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/ios/compositing/hdr/hdr-svg-inline-image-expected.txt: Added.
* LayoutTests/platform/mac-wk2/TestExpectations:
* LayoutTests/platform/mac-wk2/compositing/hdr/hdr-svg-inline-image-expected.txt: Added.
* Source/WebCore/html/ImageBitmap.cpp:
* Source/WebCore/loader/cache/CachedImage.cpp:
(WebCore::CachedImage::hasPaintedHDRContent const):
(WebCore::CachedImage::CachedImageObserver::imageContentChanged):
(WebCore::CachedImage::imageContentChanged):
(WebCore::CachedImage::headroom const): Deleted.
* Source/WebCore/loader/cache/CachedImage.h:
* Source/WebCore/loader/cache/CachedImageClient.h:
(WebCore::CachedImageClient::imageContentChanged):
* Source/WebCore/page/ChromeClient.h:
(WebCore::ChromeClient::didFinishLoadingImageForSVGImage):
* Source/WebCore/page/Page.cpp:
(WebCore::Page::didFinishLoadingImageForSVGImage):
* Source/WebCore/page/Page.h:
* Source/WebCore/platform/graphics/BitmapImage.cpp:
(WebCore::BitmapImage::draw):
* Source/WebCore/platform/graphics/BitmapImage.h:
* Source/WebCore/platform/graphics/BitmapImageSource.h:
* Source/WebCore/platform/graphics/Image.h:
(WebCore::Image::hasPaintedHDRContent const):
(WebCore::Image::headroom const): Deleted.
* Source/WebCore/platform/graphics/ImageObserver.h:
* Source/WebCore/platform/graphics/ImageSource.h:
(WebCore::ImageSource::setHasPaintedHDRContentForTesting):
(WebCore::ImageSource::hasPaintedHDRContentForTesting const):
(WebCore::ImageSource::setHeadroomForTesting): Deleted.
(WebCore::ImageSource::headroomForTesting const): Deleted.
* Source/WebCore/rendering/RenderElement.cpp:
(WebCore::RenderElement::imageContentChanged):
* Source/WebCore/rendering/RenderElement.h:
* Source/WebCore/rendering/RenderImage.cpp:
(WebCore::RenderImage::notifyFinished):
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::calculateClipRects const):
* Source/WebCore/rendering/RenderLayer.h:
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::PaintedContentsInfo::PaintedContentsInfo):
(WebCore::PaintedContentsInfo::isContentsTypeSatisfied const):
(WebCore::PaintedContentsInfo::isRenderElementWithHDR):
(WebCore::PaintedContentsInfo::determineContentsType):
(WebCore::RenderLayerBacking::updateDrawsContent):
(WebCore::RenderLayerBacking::isRenderElementWithHDR const):
(WebCore::PaintedContentsInfo::isReplacedElementWithHDR): Deleted.
(WebCore::RenderLayerBacking::isReplacedElementWithHDR const): Deleted.
* Source/WebCore/rendering/RenderLayerBacking.h:
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGImage.cpp:
(WebCore::LegacyRenderSVGImage::notifyFinished):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGImage.h:
* Source/WebCore/svg/SVGImageElement.h:
* Source/WebCore/svg/graphics/SVGImage.cpp:
(WebCore::SVGImage::hasPaintedHDRContent const):
* Source/WebCore/svg/graphics/SVGImage.h:
* Source/WebCore/svg/graphics/SVGImageClients.h:
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::setHasPaintedHDRContentForTesting):
(WebCore::Internals::setHeadroomForTesting): Deleted.
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:
* Tools/TestWebKitAPI/Tests/WebCore/SVGImageCasts.cpp:

Canonical link: https://commits.webkit.org/291982@main

smfr commented on Source/WebCore/rendering/RenderElement.cpp:
```diff
@@ -1712,6 +1712,21 @@ void RenderElement::didRemoveCachedImageClient(CachedImage& cachedImage)
+    UNUSED_PARAM(cachedImage);
+#endif
+
+    if (auto layer = layerParent())
```
This should be `enclosingLayer()`.

shallawa replied:
Fixed in https://commits.webkit.org/292035@main.
